0000-spec-distribution.md

```diff
@@ -0,0 +1,302 @@
+
+Given:
+
+<img src="http://tex.s2cms.ru/svg/%0AO%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%20%5C%5C%0AC%2C%20%26%20D%0A%5Cend%7Bbmatrix%7D%0A" alt="
+O =
+\begin{bmatrix}
+A, &amp; B \\
+C, &amp; D
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0AP%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20F%20%5C%5C%0AG%2C%20%26%20H%0A%5Cend%7Bbmatrix%7D%0A" alt="
+P =
+\begin{bmatrix}
+E, &amp; F \\
+G, &amp; H
+\end{bmatrix}
+" />
+
+We propose:
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20O%2C%20%26%20F%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20A%2C%20%26%20B%2C%20%26%20F%20%5C%5C%0AE%2C%20%26%20C%2C%20%26%20D%2C%20%26%20F%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+E, &amp; O, &amp; F
+\end{bmatrix} = 
+\begin{bmatrix}
+E, &amp; A, &amp; B, &amp; F \\
+E, &amp; C, &amp; D, &amp; F \\
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AO%2C%20%26%20P%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AA%2C%20%26%20B%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+O, &amp; P
+\end{bmatrix} = 
+\begin{bmatrix}
+A, &amp; B, &amp; E, &amp; F \\
+A, &amp; B, &amp; G, &amp; H \\
+C, &amp; D, &amp; E, &amp; F \\
+C, &amp; D, &amp; G, &amp; H \\
+\end{bmatrix}
+" />
+
+
+## User Interface
+
+### App Developer
+
+`pack build` should accept a list of buildpacks via the `--buildpack` flag.
+The flag may be passed multiple times to construct a buildpack group.
+
+The value of each flag must be one of:
+- A buildpack ID of a buildpack on the builder, optionally followed by `@` and a version
+- A path to a buildpackage on the local filesystem
+- A reference to a buildpackage on a Docker registry  
+
+A version must be specified if the version is ambiguous.
+
+### Buildpack Developer
+
+`pack create-package` will package a selection of
+
+- gzip compressed, tar archived buildpack blobs
+- other buildpackages
+- stack and buildpack metadata
+
+into a `.cnb` file, OCI image in a registry, or OCI image in a Docker daemon.
+
+These properties will be specified in a `package.toml` file.
+
+```toml
+id = "io.buildpacks.nodejs"
+version = "0.0.9"
+
+[[blobs]]
+uri = "https://example.com/nodejs.tgz"
+[[blobs.buildpacks]]
+id = "io.buildpacks.nodejs"
+version = "0.0.9"
+[[blobs.buildpacks]]
+id = "io.buildpacks.npm"
+version = "0.0.7"
+[[blobs.buildpacks]]
+id = "io.buildpacks.node"
+version = "0.0.5"
+
+[[packages]]
+ref = "registry.example.com/ruby"
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+mixins = ["build:git"]
+```
+
+`pack create-builder` will generate a builder image from buildpackages, buildpack blobs, and stack metadata.
+
+These properties will be specified in a `builder.toml` file.
+
+```toml
```

Comment:
@ekcasey @hone this UX gives builder creators maximum flexibility in defining a builder. Buildpackages are not even necessary.

Reply:
Similar feedback to above. This makes creating a builder more complicated (or maybe just confusing) given that buildpackages are now another possible component of the builder. Is the "best practice" to create a buildpackage per *the concept formerly known as a label* and then compose those or just mix and match buildpackages and buildpacks directly. As a buildpack distributor, I know have to decide that  I guess?

Reply:
A buildpackage is a collection of buildpacks with a single entrypoint ID that references all of those buildpacks (via order definitions). It may work with any number of stacks.

A builder is a single stack image with any number of buildpacks or buildpackages on it.

The builder is useful for building on a workstation, where you need a single copy of the latest buildpacks. It's also useful for building on a platform that doesn't explicitly support CNB (like tekton or concourse).

A buildpackage is more useful for distributing a group of buildpacks that could be used on CF, Heroku, or another buildpack-native platform. It's also a useful way to extend builders that might not have all the buildpacks you need.

In the long term, once we have a centralized buildpack registry that maps buildpack IDs to their package locations, I expect that the builder and package.toml concepts might fade away.